File: Solutions/581. Shortest Unsorted Continuous Subarray/Python/Shortest_Unsorted_Continuous_Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: list[int]) -> int:
        sizeof_nums = len(nums)

        find_min    = False
        minBaseLine = float('inf')
        for i in range(sizeof_nums):
            if i == sizeof_nums - 1 or nums[i + 1] < nums[i]:
                find_min = True

            if find_min:
                minBaseLine = min(minBaseLine, nums[i])

        find_max    = False
        maxBaseLine = float('-inf')
        for i in range(sizeof_nums - 1, -1, -1):
            if i == 0 or nums[i - 1] > nums[i]:
                find_max = True

            if find_max:
                maxBaseLine = max(maxBaseLine, nums[i])

        left  = -1
        for i in range(sizeof_nums):
            if (nums[i] > minBaseLine) and left == -1:
                left = i
                break

        right = -1
        for i in range(sizeof_nums - 1, -1, -1):
            if (nums[i] < maxBaseLine) and right == -1:
                right = i
                break

        return (right - left + 1) if (right - left + 1) != 1 else 0

S = Solution()

print(S.findUnsortedSubarray([1,2,3,2,3,3]))


        

File: Solutions/581. Shortest Unsorted Continuous Subarray/CPP/Shortest_Unsorted_Continuous_Subarray.cpp
#include <iostream>
#include <vector>
#include <limits>

using namespace std;

class Solution{
public:
    int findUnsortedSubarray(vector<int>& nums){
        int sizeOfnums = nums.size();

        bool find_min = false;
        int minBound = numeric_limits<int>::max();
        for(int i = 0; i < sizeOfnums; i++){
            if(find_min == true){
                minBound = min(minBound, nums[i]);
            }
            else if(i != sizeOfnums - 1 and nums[i] > nums[i + 1]){
                find_min = true;
            }
        }

        bool find_max = false;
        int maxBound = numeric_limits<int>::min();
        for(int i = sizeOfnums - 1; i >= 0; i--){
            if(find_max == true){
                maxBound = max(maxBound, nums[i]);
            }
            else if(i != 0 and nums[i - 1] > nums[i]){
                find_max = true;
            }
        }

        int left = -1;
        for(int i = 0; i < sizeOfnums; i++){
            if(nums[i] > minBound){
                left = i;
                break;
            }
        }

        int right = -1;
        for(int i = sizeOfnums - 1; i >= 0; i--){
            if(nums[i] < maxBound){
                right = i;
                break;
            }
        }

        int size = right - left + 1;
        return (size > 1) ? size : 0;
    }
};

int main(){
    Solution* S = new Solution();
    vector<int> nums ={1, 2, 3, 2, 3, 3};
    cout << S->findUnsortedSubarray(nums) << endl;
    return 0;
}